### PR TITLE
C++/C#/Java: Shared TaintTrackingImpl.qll

### PR DESCRIPTION
**Pull request by jbj**
This file is now identical in all languages. Unifying this file led to the following changes:
- The documentation spelling fixes and example from the C++ version
  were copied to the other versions and updated.
- The steps through `NonLocalJumpNode` from C# were abstracted into a
  `globalAdditionalTaintStep` predicate that's empty for C++ and Java.
- The `defaultTaintBarrier` predicate from Java is now present but empty
  on C++ and C#.
- The C++ `isAdditionalFlowStep` predicate on
  `TaintTracking::Configuration` no longer includes `localFlowStep`.
  That should avoid some unnecessary tuple copying.

### CODE REVIEW

**Comment by jbj** [comment]
I forgot to list
- The new taint wrappers for barrier guards and in/out barriers are now exposed for C# and C++.
- The `isSanitizerEdge` predicate is now deprecated for C# and C++.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll [hunk 1]
```diff
@@ -42,57 +35,79 @@ private import TaintTrackingParameter::Private
...
+  predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) { none() }
 
-  /** DEPRECATED: use `isSanitizerEdge` instead. */
-  override deprecated
-  predicate isBarrierEdge(DataFlow::Node node1, DataFlow::Node node2) {
-    this.isSanitizerEdge(node1, node2)
+  final override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    isAdditionalTaintStep(node1, node2) or
+    localAdditionalTaintStep(node1, node2) or
+    globalAdditionalTaintStep(node1, node2)
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll]: Could we replace `localAdditionalTaintStep(node1, node2) or globalAdditionalTaintStep(node1, node2)` with `defaultAdditionalTaintStep(node1, node2)` in the shared code?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll]: That's fine with me. What do you think, @hvitved ?

Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll]: Sounds fine with me.

**Comment by jbj** [comment]
I forgot the C++ IR taint tracking library, both here and in #1757. I'll fix that in a follow-up PR.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll]: Done.

**Pull request merged**